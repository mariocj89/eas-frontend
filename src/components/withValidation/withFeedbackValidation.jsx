import React from 'react';
import PropTypes from 'prop-types';

const withFeedbackValidation = WrappedComponent => {
  const WithFeedbackValidation = (props, context) => {
    const error = context.getFormError();
    return error ? <WrappedComponent error={error} /> : null;
  };

  WithFeedbackValidation.contextTypes = {
    getFormError: PropTypes.func.isRequired,
  };

  return WithFeedbackValidation;
};

export default withFeedbackValidation;
