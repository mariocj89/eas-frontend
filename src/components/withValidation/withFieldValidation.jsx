import React, { Component } from 'react';
import { translate } from 'react-i18next';

import PropTypes from 'prop-types';

const withFieldValidation = WrappedComponent => {
  class WithFieldValidation extends Component {
    constructor(props) {
      super(props);
      this.state = {
        changed: false,
        error: undefined,
      };
    }

    componentDidMount() {
      const { value, name } = this.props;
      const errors = this.getErrors(value);
      const newState = { error: errors };
      const valid = !errors;
      const isEmptyAtRegister = (Array.isArray(value) && !value.length) || value === '';
      if (!isEmptyAtRegister) {
        newState.changed = true;
      }
      this.setState(newState);
      this.context.registerValidatedField(name, valid);
    }

    componentDidUpdate(prevProps) {
      const prevErrors = this.getErrors(prevProps.value);
      const errors = this.getErrors(this.props.value);
      if (JSON.stringify(prevErrors) !== JSON.stringify(errors)) {
        this.setState({ changed: true, error: errors });
        this.context.updateFieldValidationState(this.props.name, !errors);
      }
    }

    componentWillUnmount() {
      this.context.deregisterValidatedField(this.props.name);
    }

    getDefaultErrorMessage = error => {
      const { t } = this.props;
      switch (error.rule) {
        case 'required':
          return t('default_message_required_field');
        case 'min':
          return t('default_message_min', { min: error.value });
        default:
          return '';
      }
    };

    getErrors = value => {
      const { validators } = this.props;
      return validators.find(validator => {
        switch (validator.rule) {
          case 'required':
            return !String(value).trim();
          case 'min':
            return isNaN(value) || value < parseInt(validator.value, 10);
          default:
            return true;
        }
      });
    };

    getErrorsToShow = () => {
      const { changed, error } = this.state;
      const shouldShowError = this.context.isFormSubmitted() || changed;
      return shouldShowError && error ? error : undefined;
    };

    render() {
      const { validators, t, tReady, ...props } = this.props; // eslint-disable-line react/prop-types
      const error = this.getErrorsToShow();
      let message;
      if (error) {
        message = error.message || this.getDefaultErrorMessage(error);
      }
      return (
        <WrappedComponent
          {...props}
          FormHelperTextProps={error && { 'data-has-error': true }}
          error={Boolean(error)}
          helperText={message}
        />
      );
    }
  }

  WithFieldValidation.propTypes = {
    onChange: PropTypes.func,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.arrayOf(PropTypes.string),
    ]),
    name: PropTypes.string.isRequired,
    validators: PropTypes.arrayOf(
      PropTypes.shape({
        rule: PropTypes.string.isRequired,
        value: PropTypes.any,
        message: PropTypes.string,
      }),
    ),
    t: PropTypes.func.isRequired,
  };

  WithFieldValidation.defaultProps = {
    onChange: null,
    validators: [],
    value: '',
  };

  WithFieldValidation.contextTypes = {
    registerValidatedField: PropTypes.func.isRequired,
    deregisterValidatedField: PropTypes.func.isRequired,
    updateFieldValidationState: PropTypes.func.isRequired,
    isFormSubmitted: PropTypes.func.isRequired,
  };

  return translate('WithFieldValidation')(WithFieldValidation);
};

export default withFieldValidation;
