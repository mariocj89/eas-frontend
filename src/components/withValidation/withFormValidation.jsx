import React, { Component } from 'react';
import PropTypes from 'prop-types';

const withFormValidation = WrappedComponent => {
  class WithFormValidation extends Component {
    constructor(props) {
      super(props);
      this.state = {
        formSubmitted: false, // It's used to bypass changedFields when the form is submitted
        validations: {},
      };
    }

    getChildContext() {
      return {
        registerValidatedField: this.registerValidatedField.bind(this),
        deregisterValidatedField: this.deregisterValidatedField.bind(this),
        getFormError: this.getFormError.bind(this),
        updateFieldValidationState: this.updateFieldValidationState.bind(this),
        isFormSubmitted: this.isFormSubmitted.bind(this),
      };
    }

    onSubmit = e => {
      if (!this.isFormValid()) {
        e.preventDefault();
        this.setState({
          formSubmitted: true,
        });
      } else {
        this.props.onSubmit(e);
      }
    };

    getFormError() {
      const { formSubmitted } = this.state;
      console.log(
        'formSubmitted ? this.props.checkErrors()',
        formSubmitted,
        this.props.checkErrors(),
      );
      return formSubmitted ? this.props.checkErrors() : undefined;
    }

    isFormValid = () => {
      const validations = Object.values(this.state.validations);
      const fieldsValid = validations.every(Boolean);
      const formValid = !this.props.checkErrors();
      return fieldsValid && formValid;
    };

    isFormSubmitted = () => this.state.formSubmitted;

    registerValidatedField(name, valid) {
      this.updateFieldValidationState(name, valid);
    }

    updateFieldValidationState(name, valid) {
      this.setState(
        previousState => ({
          validations: {
            ...previousState.validations,
            ...{
              [name]: valid,
            },
          },
        }),
        () => {
          this.triggerValidationChange();
        },
      );
    }

    triggerValidationChange = () => {
      if (this.props.onValidationChange) {
        const isValid = this.isFormValid();
        this.props.onValidationChange(isValid);
      }
    };

    deregisterValidatedField(name) {
      this.setState(
        previousState => {
          const validations = Object.assign({}, previousState.validations);
          delete validations[name];
          return {
            validations,
          };
        },
        () => {
          if (this.props.onFieldDeregister) {
            this.props.onFieldDeregister(name);
          }
          this.triggerValidationChange();
        },
      );
    }

    render() {
      const { onSubmit, onFieldDeregister, onValidationChange, checkErrors, ...props } = this.props;
      return <WrappedComponent {...props} onSubmit={this.onSubmit} />;
    }
  }

  WithFormValidation.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    onFieldDeregister: PropTypes.func,
    onValidationChange: PropTypes.func,
    checkErrors: PropTypes.func,
  };

  WithFormValidation.defaultProps = {
    onFieldDeregister: null,
    onValidationChange: null,
    checkErrors: () => {},
  };

  WithFormValidation.childContextTypes = {
    registerValidatedField: PropTypes.func,
    deregisterValidatedField: PropTypes.func,
    getFormError: PropTypes.func,
    updateFieldValidationState: PropTypes.func,
    isFormSubmitted: PropTypes.func,
  };

  return WithFormValidation;
};

export default withFormValidation;
