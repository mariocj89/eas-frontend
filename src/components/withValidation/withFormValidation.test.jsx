import React from 'react';
import PropTypes from 'prop-types';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import withFormValidation from './withFormValidation';
import withFieldValidation from './withFieldValidation';
// import ValidationFeedback from './ValidationFeedback';
import withFeedbackValidation from './withFeedbackValidation';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  translate: () => Component => {
    // eslint-disable-next-line no-param-reassign
    Component.defaultProps = { ...Component.defaultProps, t: () => 'some translated stuff' };
    return Component;
  },
}));

// eslint-disable-next-line
const Form = ({ onFieldRegister, onFieldDeregister, onFieldChange, ...props }) => (
  <form {...props} />
);

const simulateChangeField = (element, value) => {
  const event = { target: { name: element.props().name, value } };
  element.simulate('change', event);
};

// eslint-disable-next-line
const Field = ({ valid, FormHelperTextProps, helperText, error, ...props }) => <input type="text" {...props} />;

const FormWithValidation = withFormValidation(Form);
const FieldwithValidation = withFieldValidation(Field);
const ValidationFeedback = withFeedbackValidation(() => <div />);

describe('withFormValidation', () => {
  it('should render correctly', () => {
    const wrapper = mount(<FormWithValidation onSubmit={jest.fn()} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('Registering fields', () => {
    it('should register validated fields', () => {
      const wrapper = mount(<FormWithValidation onSubmit={jest.fn()} />);
      const instance = wrapper.instance();

      instance.registerValidatedField('field1', true);
      instance.registerValidatedField('field2', false);

      wrapper.update();

      const { state } = instance;

      expect('field1' in state.validations).toBe(true);
      expect('field2' in state.validations).toBe(true);
      expect(state.validations.field1).toBe(true);
      expect(state.validations.field2).toBe(false);
    });

    it('should deregister validated fields', () => {
      const wrapper = mount(<FormWithValidation onSubmit={jest.fn()} />);
      const instance = wrapper.instance();

      instance.registerValidatedField('field1', false);
      instance.registerValidatedField('field2', true);

      instance.deregisterValidatedField('field2');
      wrapper.update();

      const { state } = instance;

      expect('field1' in state.validations).toBe(true);
      expect('field2' in state.validations).toBe(false);
    });
  });

  describe('Form errors', () => {
    it.only('Should not be shown if form was not submmited', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => 'There are errors'}>
          <ValidationFeedback />
        </FormWithValidation>,
      );
      const instance = wrapper.instance();
      instance.setState({ formSubmitted: true });
      wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should be shown if when form is submitted', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => 'There are errors'}>
          <ValidationFeedback />
        </FormWithValidation>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Are not shown if there are not errors', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => undefined}>
          <FieldwithValidation name="field1" id="field" value="1" />
          <ValidationFeedback />
        </FormWithValidation>,
      );
      const instance = wrapper.instance();
      expect(instance.isFormValid()).toBe(true);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Are not shown until a field is changed', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => 'There are errors'}>
          <FieldwithValidation name="field1" id="field" value="1" />
          <ValidationFeedback />
        </FormWithValidation>,
      );
      const instance = wrapper.instance();
      expect(instance.isFormValid()).toBe(false);

      const fieldDomElement = wrapper.find('#field').last();
      simulateChangeField(fieldDomElement, 'still invalid');
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
describe('Submit', () => {
  it('Should call the main onSubmit function if the form is valid', () => {
    const onSubmitMock = jest.fn();
    const wrapper = mount(<FormWithValidation onSubmit={onSubmitMock} />);

    const instance = wrapper.instance();
    instance.isFormValid = jest.fn(() => true);

    wrapper.find('form').simulate('submit');
    expect(onSubmitMock).toHaveBeenCalled();
  });

  it('Should not call the main onSubmit function if the form is invalid', () => {
    const onSubmitMock = jest.fn();
    const wrapper = mount(<FormWithValidation onSubmit={onSubmitMock} />);

    const instance = wrapper.instance();
    instance.isFormValid = jest.fn(() => false);

    wrapper.find('form').simulate('submit');
    expect(onSubmitMock).not.toHaveBeenCalled();
  });
});
