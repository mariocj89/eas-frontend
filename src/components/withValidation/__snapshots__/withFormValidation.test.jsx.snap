// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFormValidation Form errors Are not shown if there are not errors 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[Function]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <WithFieldValidation
        id="field"
        name="field1"
        onChange={null}
        t={[Function]}
        validators={Array []}
        value="1"
      >
        <Field
          error={false}
          id="field"
          name="field1"
          onChange={null}
          value="1"
        >
          <input
            id="field"
            name="field1"
            onChange={null}
            type="text"
            value="1"
          />
        </Field>
      </WithFieldValidation>
      <WithFeedbackValidation>
        <ValidationFeedback
          error=""
        />
      </WithFeedbackValidation>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation Form errors Are not shown until a field is changed 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[Function]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <WithFieldValidation
        id="field"
        name="field1"
        onChange={null}
        t={[Function]}
        validators={Array []}
        value="1"
      >
        <Field
          error={false}
          id="field"
          name="field1"
          onChange={null}
          value="1"
        >
          <input
            id="field"
            name="field1"
            onChange={null}
            type="text"
            value="1"
          />
        </Field>
      </WithFieldValidation>
      <WithFeedbackValidation>
        <ValidationFeedback
          error=""
        />
      </WithFeedbackValidation>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation Form errors Should be shown if when form is submitted 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[Function]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <WithFeedbackValidation>
        <ValidationFeedback
          error=""
        />
      </WithFeedbackValidation>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation Form errors Should not be shown if form was not submmited 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[Function]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <WithFeedbackValidation>
        <ValidationFeedback
          error="There are errors"
        >
          <div
            className="ValidationFeedback"
            data-component="ValidationFeedback"
          >
            error
            <BannerAlert
              title="There are errors"
              type="error"
            >
              <section
                className="BannerAlert BannerAlert--error"
              >
                <pure(Info)
                  className="BannerAlert__error-icon"
                  style={
                    Object {
                      "fontSize": 18,
                    }
                  }
                >
                  <Info
                    className="BannerAlert__error-icon"
                    style={
                      Object {
                        "fontSize": 18,
                      }
                    }
                  >
                    <WithStyles(SvgIcon)
                      className="BannerAlert__error-icon"
                      style={
                        Object {
                          "fontSize": 18,
                        }
                      }
                    >
                      <SvgIcon
                        className="BannerAlert__error-icon"
                        classes={
                          Object {
                            "colorAction": "MuiSvgIcon-colorAction-4",
                            "colorDisabled": "MuiSvgIcon-colorDisabled-6",
                            "colorError": "MuiSvgIcon-colorError-5",
                            "colorPrimary": "MuiSvgIcon-colorPrimary-2",
                            "colorSecondary": "MuiSvgIcon-colorSecondary-3",
                            "fontSizeInherit": "MuiSvgIcon-fontSizeInherit-7",
                            "root": "MuiSvgIcon-root-1",
                          }
                        }
                        color="inherit"
                        component="svg"
                        fontSize="default"
                        style={
                          Object {
                            "fontSize": 18,
                          }
                        }
                        viewBox="0 0 24 24"
                      >
                        <svg
                          aria-hidden="true"
                          className="MuiSvgIcon-root-1 BannerAlert__error-icon"
                          focusable="false"
                          style={
                            Object {
                              "fontSize": 18,
                            }
                          }
                          viewBox="0 0 24 24"
                        >
                          <g>
                            <path
                              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                            />
                          </g>
                        </svg>
                      </SvgIcon>
                    </WithStyles(SvgIcon)>
                  </Info>
                </pure(Info)>
                 
                <span>
                  There are errors
                </span>
                 
              </section>
            </BannerAlert>
          </div>
        </ValidationFeedback>
      </WithFeedbackValidation>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation should render correctly 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[Function]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    />
  </Form>
</WithFormValidation>
`;
